File: 2020/4/a/a.go
package main

import (
	"2020/common"
	"log"
	"strings"
)

var mandatoryKeys = []string{
	"byr",
	"iyr",
	"eyr",
	"hgt",
	"hcl",
	"ecl",
	"pid",
}

var notMandatoryKeys = []string{
	"cid",
}

func main() {
	file := common.OpenInputFile("4/input.txt")
	buffer := common.ReadBuffer(file)
	var validPassports = 0
	passportKeys := make([]string, 0)
	for buffer.Scan() {
		line := string(buffer.Bytes())
		if line == "" {
			if isPassportValid(passportKeys) {
				validPassports++
			}
			passportKeys = make([]string, 0)
			continue
		}
		tokens := strings.Split(line, " ")
		for _, token := range tokens {
			key := strings.Split(token,":")
			passportKeys = append(passportKeys, key[0])
		}
	}
	// Handle EOF
	if isPassportValid(passportKeys) {
		validPassports++
	}

	if err := buffer.Err(); err != nil {
		log.Fatal(err)
	}
	log.Printf("Total valid passports: %d", validPassports)
}

func isPassportValid(keys []string) bool {
	count := 0
	for _, key := range keys {
		for _, mandatory := range mandatoryKeys {
			if key == mandatory {
				count++
			}
		}
	}
	if count == len(mandatoryKeys) {
		return true
	}
	return false
}

File: 2020/4/a/a.rs
fn main() {
    a();
}

fn a() {
    let mandatory_keys: Vec<&str> = vec!["byr",
                                            "iyr",
                                            "eyr",
                                            "hgt",
                                            "hcl",
                                            "ecl",
                                            "pid"
    ];

    let list = advent_of_code::file::read_stream();
    let mut store_passport: Vec<String> = Vec::new();
    let mut count_valid_passport = 0;
    for line in list {
        let unwrapped_line = line.unwrap();
        if unwrapped_line == "" {
            if is_valid_passport(store_passport, mandatory_keys.clone()) {
                count_valid_passport += 1;
            }
            store_passport = Vec::new();
            continue;
        }
        let elts = unwrapped_line.split_whitespace();
        for elt in elts {
            let split: Vec<&str> = elt.split(":").collect();
            store_passport.push(split[0].to_string())
        }
    }
    if is_valid_passport(store_passport, mandatory_keys.clone()) {
        count_valid_passport += 1;
    }
    println!("The value is: {:#?}", count_valid_passport);
}

fn is_valid_passport(keys: Vec<String>, mandatory_keys: Vec<&str>) -> bool {
    let mut counter = 0;
    for key in &mandatory_keys {
        if keys.contains(&key.to_string()) {
            counter += 1
        }
    }
    if counter == mandatory_keys.len() {
        return true
    }
    return false
}